### DISCUSSION TITLE
$apis.recordAuthResponse seem to break the route

### DISCUSSION THREAD

**Discussion by square-foot** [pocketbase/pocketbase#3330] [Q&A] [answered]
I created a route in the GoJa server side Javascript. I believe by default, it does not check the api_rules on any collection (pls correct me if I am wrong) So I used this logic: Artificially insert the user's record into the 'c' context (as I can send the user's ID xxxxxx into the route ... and then later on check if the records I want are accessible to the user ... but see problem I encountered explained after this code:

    routerAdd("GET", "/opdsessions", (c) => {
        //... code not shown 
    
        ///artificially insert the record of the user into  'c'  
         const recordusr = $app.dao().findFirstRecordByData("users", "id", "xxxxxxxx")  
     
         $apis.recordAuthResponse($app, c, recordusr)
    
              const info   = $apis.requestInfo(c);
            
        ///use the info object to check if the user is allowed to access a collection named "opdsessions"

         const records = $app.dao().findRecordsByFilter("opdsessions", "1 = 1","-sessionstart",100)

        records.forEach((rec)=>{

              //CHECK if each record is actually usable by user!
              
              const canAccess = $app.dao().canAccessRecord(rec, info, rec.collection().viewRule)
              if (!canAccess) {
                  throw new ForbiddenError()
                }

         })

         return c.html(200,"Done")

         }
 
   
What I find is that the route returns the recordusr back to the browser for that route after the _$apis.recordAuthResponse(...)_ call . So the c.html(200,"Done") does not even come even if the user was permitted.
   
I got a workaround for this.  This is what I did:
  
      routerAdd("GET", "/opdsessions", (c) => {
        //... code not shown 
    
  
         const recordusr = $app.dao().findFirstRecordByData("users", "id", "xxxxxxxx")  
     
         //NOTE: do not use $apis.recordAuthResponse(...)
         //$apis.recordAuthResponse($app, c, recordusr)
    
            //create an "info" object of my own
              const info   = {authRecord: recordusr}
            
        ///use the info object to check if the user is allowed to access a collection named "opdsessions"

         const records = $app.dao().findRecordsByFilter("opdsessions", "1 = 1","-sessionstart",100)

        records.forEach((rec)=>{

              //CHECK if each record is actually usable by user!
              
              const canAccess = $app.dao().canAccessRecord(rec, info, rec.collection().viewRule)
              if (!canAccess) {
                  throw new ForbiddenError()
                }

         })

         return c.html(200,"Done")

         }
 
  
 Not sure if I am understanding this correctly.  But the second example I gave above is working. First one is not.

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand the above code samples and what you are trying to do.

`$apis.recordAuthResponse()` writes a formatted auth response, it doesn't _"artificially insert the record of the user into  'c'"_

After this call the response of the body is already written/committed and any consequent response body write calls will not have effect.

**Reply by square-foot** [reply]
Ah... my misunderstanding. I thought  `$apis.recordAuthResponse` was for making sure that 'c' gets the authinfo.  

If I use  `const info   = $apis.requestInfo(c); `  it always gives `{"code":403,"message":"You are not allowed to perform this request.","data":{}}` even if my browser has logged in the correct user 

So I assumed that the Goja server code has no clue about who had logged in when that route was called in that session. I guess I need to pass some headers for such calls?

**Reply by ganigeorgiev** [reply, marked as answer]
I'm sorry, I'm still failing to understand what you are trying to do.

> I guess I need to pass some headers for such calls?

Yes, the `Authorization: TOKEN` header.
PocketBase APIs are stateless and we don't use session cookies or similar. Each request must have a valid `Authorization` header in order an auth record (or admin) to be considered authenticated.

**Reply by square-foot** [reply]
Thank you. That clarifies a lot. 

I am just trying out various scenarios by which I can output records from collections to the browser. Eventually, my goal is to create some kind of boilerplate/code-generator to do some common activities with Pocketbase. I am still learning. It is an excellent system
